label-application-rule-config-service-api/src/main/proto/org/hypertrace/label/application/rule/config/service/v1/label_application_rule.proto

```diff
@@ -0,0 +1,98 @@
+syntax = "proto3";
+
+package org.hypertrace.label.application.rule.config.service.v1;
+
+option java_multiple_files = true;
+
+message LabelApplicationRule {
+  string id = 1;
+  LabelApplicationRule data = 2;
+}
+
+message CreateLabelApplicationRule {
+  LabelApplicationRule data = 2;
+}
+
+message LabelApplicationRuleData {
+  string name = 1;
+  string description = 2;
+  MatchingCriteria matching_criteria = 3;
+  ApplicationCriteria application_criteria = 4;
+
+  message MatchingCriteria {
+    Type type = 1;
+    repeated Condition conditions = 2;
+
+    enum Type {
+      TYPE_UNSPECIFIED = 0;
+      TYPE_SPAN = 1;
+      TYPE_TRACE = 2;
+    }
+
+    message Condition {
+      string name = 1;
+      Scope scope = 2;
+      string key = 3;
+      Operator operator = 4;
+      optional string value = 5;
+
+      enum Scope {
+        SCOPE_UNSPECIFIED = 0;
+        SCOPE_ATTRIBUTE = 1;
+        SCOPE_REQUEST = 2;
+        SCOPE_RESPONSE = 3;
+      }
+
+      enum Operator {
+        OPERATOR_UNSPECIFIED = 0;
+        OPERATOR_EXISTS = 1;
+        OPERATOR_EQUALS = 2;
+        OPERATOR_CONTAINS = 3;
+        OPERATOR_STARTS_WITH = 4;
+        OPERATOR_ENDS_WITH = 5;
+        OPERATOR_MATCHES_REGEX = 6;
+      }
+    }
+  }
+
+  message ApplicationCriteria {
+    Type type = 1;
+    Scope scope = 2;
+    Action action = 3;
+
+    enum Type {
+      TYPE_UNSPECIFIED = 0;
+      TYPE_ENTITY = 1;
+    }
+
+    enum Scope {
+      SCOPE_UNSPECIFIED = 0;
+      SCOPE_API = 1;
+      SCOPE_SERVICE = 2;
+      SCOPE_DOMAIN = 3;
+    }
+
+    message Action {
+      Operation operation = 1;
+      oneof value {
+        string static_label_id = 2;
+        DynamicLabel dynamic_label = 3;
+      }
+
+      enum Operation {
+        OPERATION_UNSPECIFIED = 0;
+        OPERATION_MERGE = 1;
+      }
+
+      message DynamicLabel {
+        repeated TokenExtractionRule token_extraction_rules = 1;
+        string token_combining_expression  = 2;
+
+        message TokenExtractionRule {
+          string key = 1;
```

Comment:
Are these span keys? trace keys?

Reply:
Will be picked from type defined in matching criteria?

Reply:
I think this is where part of the confusion of scope comes in, too. What's being called type in matching criteria is scope elsewhere in the system. "API", "SERVICE", "SPAN", "TRACE" - all scopes.

Further, we're designed to enforce a single scope across all conditions. Do we think it will ever be a requirement to - for example - want a trace to match condition X and a span to match condition Y? I could definitely envision that, as the trace gives data from an end user perspective (e.g. what was the host header on the external call), while the span gives more specific data (what service is this, what data was passed in etc.).

Reply:
I have defined type and scope. Type can be `span or trace` for matching criteria, and `entity` for application criteria. Source can be `request, response or attribute` for key. If this terminology is confusing, then we can change it.

Reply:
There may be cases where heterogeneous conditions support is required.  To support  heterogeneous conditions, we can have `type` at `condition level`. Similarly we will need to expose `type` in `token extraction rule` as well.

Reply:
I think potentially prefixing the words like `target_scope`, `matching_scope` etc. would make sense if all of the values are subsets of our general scope (which span, trace and entities are). WRT request/response/attribute, I'd use a different word there if we preserve it, but I'd hope to get away from it as discussed on a different thread.